common/src/stack/pylib/stack/kvm.py

```diff
@@ -0,0 +1,376 @@
+			pool.undefine()
+		except libvirtError as msg:
+			raise VmException(f'Failed to delete pool {poolname} on hypervisor {self.hostname}:\n{msg}')
+
+	def add_volume(self, volname, pooldir, poolname, size):
+		"""
+		Add a new storage volume to a pool
+
+		Raises a VmException if the volume couldn't be added
+		"""
+
+		volxml = f"""<volume>
+	<name>{volname}</name>
+	<allocation>0</allocation>
+	<capacity unit="G">{size}</capacity>
+	<target>
+		<path>{pooldir}/{volname}</path>
+		<format type="qcow2"/>
+	</target>
+	</volume>"""
+
+		error_msg = f'Failed to create volume {volname} on hypervisor {self.hostname}'
+
+		try:
+			pool = self.kvm.storagePoolLookupByName(poolname)
+		except libvirtError as msg:
+			raise VmException(f'{error_msg}:\n{msg}')
+
+			# If there is a volume in the same pool matching the name
+			# parameter assume that this means the volume is already created
+		try:
+			vol = pool.storageVolLookupByName(volname)
+			return
+		except libvirtError:
+			pass
+		try:
+			pool.createXML(volxml, 0)
+		except libvirtError as msg:
+			raise VmException(f'{error_msg}:\n{msg}')
+
+	def remove_volume(self, poolname, volname):
+		"""
+		Remove a storage volume from a pool
+
+		Raises a VmException if the volume
+		failed to be deleted
+		"""
+
+		try:
+			pool = self.kvm.storagePoolLookupByName(poolname)
+			vol = pool.storageVolLookupByName(volname)
+			vol.delete()
+		except libvirtError as msg:
+			raise VmException(f'Failed to delete volume {volname} on hypervisor {self.hostname}:\n{msg}')
+
+	def copy_image(self, image_path, dest_path, image_name = ''):
+		if Path(image_path).is_file():
+			image = Path(image_path)
+			dest = Path(dest_path)
+			curr_image = Path(f'{dest}/{image_name}')
+			copy_image = f'{dest}/{image.name}'
+
+			# Create the path on the host
+			create_dest = _exec(f'ssh {self.hostname} "mkdir -p {dest}"', shlexsplit=True)
+			if create_dest.returncode != 0:
+				raise VmException(f'Could not create image folder {dest} on {self.hostname}:\n{create_dest.stderr}')
+
+			# Check if the image already exists at the given location
+			if image_name:
+				image_present = _exec(f'ssh {self.hostname} "ls {curr_image}"', shlexsplit=True)
+				if image_present.returncode == 0:
+					return
+
+			# Transfer the image
+			transfer_image = _exec(f'scp {image} {self.hostname}:{dest}', shlexsplit=True)
+			if transfer_image.returncode != 0:
+				raise VmException(f'Failed to transfer image {image} to {self.hostname} at {image}:\n{transfer_image.stderr}')
+                        # Use tar to uncompress the image if its in a tarfile
+			if tarfile.is_tarfile(image_path):
+				untar = _exec(f'ssh {self.hostname} "tar -xvf {copy_image} -C {dest} && rm {copy_image}"', shlexsplit=True)
+				if untar.returncode != 0:
+					raise VmException(f'Failed to unpack vm image {image} on {self.hostname}:\n{untar.stderr}')
+
+			# Otherwise use gunzip if its compressed using gzip
+			elif 'gz' in image.name:
```

Comment:
This would match if `gz` is anywhere in the image name. Safer to do `image.name.endswith('.gz')`